arm: nxp: update flash controller to use device tree

galak opened the pull request:
1. get flash label for k64 flash controller from dt
2. updated flash nodes to set write-block-size on all K-series NXP devices

galak committed:
drivers: flash: NXP k6x/kw2xd: Convert to use device tree

Convert NXP k6x and kw2xd flash driver to use device tree to get the
flash controller name from device tree.  We introduce yaml bindings for
the "nxp,kinetis-ftfe" and "nxp,kinetis-ftfl" devices.

Fixes: #5788

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
dts: flash: arm: nxp: Add write-block-size properties

On the various NXP Kinetis SoCs add the write-block-size property and
set it to match FSL_FEATURE_FLASH_PFLASH_BLOCK_WRITE_UNIT_SIZE for the
given SoC.

Fixes: #5788

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

jfischer-phytec-iot commented on dts/bindings/flash_controller/nxp,k64f-flash-controller.yaml:
```diff
@@ -0,0 +1,19 @@
+---
```
I guess you can rename it to nxp,kinetis-flash-controller.yaml

MaureenHelm commented on dts/bindings/flash_controller/nxp,k64f-flash-controller.yaml:
```diff
@@ -0,0 +1,19 @@
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "nxp,k64f-flash-controller"
```
This flash controller is used across the Kinetis family. Let's change the constraint to `nxp,kinetis-flash-controller` so we don't have k64 references in kw2xd, etc.

galak replied:
> This flash controller is used across the Kinetis family. Let's change the constraint to nxp,kinetis-flash-controller so we don't have k64 references in kw2xd, etc.

So looking at the register map the KW40/KW41/KL25 seem to be a bit different from K64/KW2XD.  So wondering if there should be 2 different compatibles, one for k64/kw2xd and one for kw40/kw41/kl25.

MaureenHelm replied:
You're right. The mcux flash driver abstracts the hardware differences so I mistakenly thought the flash controller was the same. Looking more closely at the register maps and hardware reference manuals, there should be three different compatibles:
nxp,kinetis-ftfa (kl25, kw40, kw41)
nxp,kinetis-ftfe (k64)
nxp,kinetis-ftfl (kw2xd)

galak replied:
Doe the 'a' 'e' 'l' have any meaning?

MaureenHelm replied:
The 'a' 'e' 'l' variants are selected based on the series (Kinetis L-series vs. K-series) and the size of the flash (greater than or less than 512 KB). I haven't been able to track down an explanation of why those particular letters or if they are an abbreviation for something.
https://www.nxp.com/docs/en/application-note/AN4835.pdf

galak commented:
Pushed an update, want to make sure this looks good on the yaml & what not for ftfe.  Than will add ftfa & ftfl yaml, dts, updates. etc.

MaureenHelm commented on arch/arm/soc/nxp_kinetis/kwx/dts.fixup:
```diff
@@ -0,0 +1,8 @@
+/* SoC level DTS fixup file */
+
+#if defined(CONFIG_SOC_MKW22D5) || defined(CONFIG_SOC_MKW24D5)
+#define FLASH_DEV_BASE_ADDRESS		NXP_K64F_FLASH_CONTROLLER_4001F000_BASE_ADDRESS_0
```
This needs to be updated to FTFL

galak replied:
Yeah, will fix that up when I introduce FTFL.

MaureenHelm commented on dts/arm/nxp/nxp_kw2xd.dtsi:
```diff
@@ -57,8 +57,9 @@
-			reg = <0x4001f000 0x27c>;
+			compatible = "nxp,kinetis-ftfl";
+			label = "FLASH_CTRL";
+			reg = <0x4001f000 0x240>;
```
Compatible looks good, let's also update the base address to the FTFL_BASE=0x40020000 rather than the FMC_BASE. You already made the same change for k6x

galak replied:
Agreed.

galak committed:
drivers: flash: NXP KL2X/KW4xZ: Add device tree support

Add device tree support for the "nxp,kinetis-ftfa" flash controller used
on the NXP KL2X and KW4xZ SoCs.

Fixes: #5788

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
@MaureenHelm should be updated now for all FTF-A,L,E and associated SoCs.
